### PR TITLE
feat: Template field workflow

### PR DESCRIPTION
**Pull request by elzody**
* Resolves: [nextcloud-gmbh/office-team#51](https://github.com/nextcloud-gmbh/office-team/issues/51)

## Summary

This PR introduces the server API and front end changes for filling out template form controls before creating a document based on a template which has fields. If a template doesn't have fields, it will continue like the normal process before.

## TODO

- [ ] Testing -- test coverage will likely be implemented in a follow-up PR
- [ ] Documentation -- if not updated, will also be updated in a follow-up PR

## Checklist

- [x] Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting) 
- [x] [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Screenshots before/after for front-end changes

### CODE REVIEW

File: lib/public/Files/Template/Template.php [hunk 1]
```diff
@@ -48,6 +50,13 @@ public function setHasPreview(bool $hasPreview): void {
 		$this->hasPreview = $hasPreview;
 	}
 
+	/**
+	 * @since 30.0.0
+	 */
+	public function setFields(array $fields): void {
```
Comment (juliushaertl) [inline lib/public/Files/Template/Template.php:56]: Maybe we could actually move the Field class type to server and make it a public API.

File: lib/public/Files/Template/Field.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Richdocuments\Template;
+
+class Field {
+	private FieldType $type;
+	private int $index;
```
Comment (juliushaertl) [inline lib/public/Files/Template/Field.php:12]: Maybe we make this a string as identifier to be more generic to also be able to identify b a unique id rather then an index

**Review by juliusknorr — APPROVED** [review]

**Comment by skjnldsv** [comment]
> * test coverage will ~~likely~~ be implemented in a follow-up PR


Would be really preferable :pray:

File: apps/files/src/components/TemplateFiller.vue [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+			this.localFields[index] = {
+				content: value,
+			}
+		},
+		async submit() {
+			this.loading = true
+
+			await this.onSubmit(this.localFields)
+
+			this.show = false
```
Comment (skjnldsv) [inline apps/files/src/components/TemplateFiller.vue:79]: Wouldn't that leave the modal instance alive?
Emitting a close event should be enough

Comment (juliushaertl) [reply apps/files/src/components/TemplateFiller.vue]: Fixed

File: lib/public/Files/Template/Field.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	 */
+	public function __construct(string $index, string $content, string|FieldType $type, ?string $alias = null, ?int $id = null, ?string $tag = null) {
+		$this->index = $index;
+		$this->alias = $alias;
+		$this->id = $id;
+		$this->tag = $tag;
+		$this->content = $content;
+
+		if ($type instanceof FieldType) {
+			$this->type = $type;
```
Comment (blizzz) [inline lib/public/Files/Template/Field.php:34]: why is `$type` allowed to be a string then?

**Review by blizzz — COMMENTED** [review]
one question, skipped looking at vue files.

File: lib/public/Files/Template/Field.php (hunk 4 above)
Comment (juliushaertl) [reply lib/public/Files/Template/Field.php]: Adjusted

Comment (juliushaertl) [reply lib/public/Files/Template/Field.php]: And moved to https://github.com/nextcloud/richdocuments/pull/3798/commits/5f730415c096a4201011e74e51490ec934808a10

**Review by skjnldsv — APPROVED** [review]

**Review by blizzz — APPROVED** [review]

**Pull request merged**